src/main/java/nextstep/qna/service/QnAService.java

```diff
@@ -25,25 +24,15 @@ public class QnAService {
 
     @Transactional
     public void deleteQuestion(NsUser loginUser, long questionId) throws CannotDeleteException {
-        Question question = questionRepository.findById(questionId).orElseThrow(NotFoundException::new);
-        if (!question.isOwner(loginUser)) {
-            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
-        }
-
-        List<Answer> answers = question.getAnswers();
-        for (Answer answer : answers) {
-            if (!answer.isOwner(loginUser)) {
-                throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
-            }
-        }
+        Question question = findQuestion(questionId);
+        question.checkIfQuestionOwner(loginUser);
+        question.checkIfAnswerOwner(loginUser);
+        question.setDeleted();
+        deleteHistoryService.saveAll(question.generateDeleteHistories());
```

Comment:
깔끔하게 리팩토링 하셨네요 👍 
다만 check~ 메서드를 모두 호출을 해야한다는 사실을 모른다면 일부 조건만 충족하더라도 삭제가 가능한 상황이 됩니다.
도메인 객체가 스스로 판단해서 삭제 결과를 리턴하면 어떨까요?
```suggestion
        Question question = findQuestion(questionId);
        List<DeleteHistory> histories = question.delete(loginUser);
        deleteHistoryService.saveAll(histories);
```

Reply:
제안해주신 코드가 훨씬 깔끔하네요.
반영하겠습니다.